src/datasets/iterable_dataset.py

```diff
@@ -2396,15 +2514,19 @@ def map(
             drop_last_batch=drop_last_batch,
             remove_columns=remove_columns,
             fn_kwargs=fn_kwargs,
-            formatting=self._formatting,
+            # pass formatting bc we need to know that we can call iter_arrow if the formatting is arrow
+            formatting=copy.deepcopy(self._formatting)
+            if self._formatting and self._formatting.format_type == "arrow"
+            else None,  # formatting is handled within ex_iterable
+            return_features=features,
         )
         info = self.info.copy()
         info.features = features
         return IterableDataset(
             ex_iterable=ex_iterable,
             info=info,
             split=self._split,
-            formatting=self._formatting,
+            formatting=None,  # formatting is applied in mapped ex_iterable; no need to re-apply it in IterableDataset
```

Comment:
you mean formatting is already applied to the input, but not to the output afaik ?

Reply:
yes - i guess what i was thinking is that if a user calls

```
ds.with_format(format_type).map(function)
```

then they are responsible for choosing the format applied to the mapped function outputs:

if the mapped function does nothing, then the formatting will be format_type.

it doesn't seem ideal to re-format the users outputs (providing no way of overriding this).

e.g if i do with_format("numpy").map(fn) where fn extracts some python object from numpy arrays, I don't want to call np.asarray(obj) on the resulting object

Reply:
> it doesn't seem ideal to re-format the user's outputs (providing them with no way of overriding this).
> e.g if i do with_format("numpy").map(fn) where fn extracts some python object from numpy arrays, I don't want to have np.asarray(obj) being called on the resulting object by the numpy formatter

You can keep in `self._formatting` because `np.asarray(obj)` is cheap on an object that is already a numpy array (zero copy). And if someone does this
```python
ds.with_format(format_type).map(function).with_format(another_type)
```
then np.asarray would not be called on the output since we can just replace the `IterableDataset._formatting` with the new one

Reply:
I think my question around this was that there is no way to just prevent any formatting happening on the outputs of map, because the user can't set _formatting to None (not what with_format(None) does).

So if map already takes care of formatting it risks introducing an expensive re-formatting step.

Reply:
Maybe `with_format(None)` should imply no formatting (contrary to the current logic which is python formatting) ? would that work ?

Reply:
I think so!

Reply:
ok let's do this then, I think it should do the job ! feel free to ping me if there is anything blocking

Reply:
Kind of still feel like setting formatting to None should be default behaviour for map? it's such an easy trap to fall into and can be expensive with any kind of complex transformation in map

Reply:
> On reflection, kind of still feel like setting formatting to None should be default behaviour for map? it's such an easy trap to fall into to not do it and it can be expensive with any kind of complex transformation in map

It would make it inconsistent with `Dataset.map` unfortunately :/
Also I've rarely seen people going from one format to another between maps

Reply:
Looking at the Dataset.map code, is it right that formatting doesn't get re-applied to output of map?

Preventing that could be another way to fix the iterable dataset case? (with_format None also seems fine but affects a bit more things which makes me slightly nervous...but can do what you think best?)

Reply:
@lhoestq what is the intended chaining logic? Should with_format.map.with_format override the original with_format and change the formatting of the inputs to map? Or is it intended to apply to the outputs?

Realise I haven't been clear about this in my head so far, and know your comment above was somewhat related but wanted to be sure I understand!

Reply:
No problem, this is not as easy as it seems though you've been doing great ! Anyway, to answer your questions:

> is it right that formatting doesn't get re-applied to output of Dataset.map?

MappedExamplesIterable doesn't apply formatting, but the IterableDataset does apply formatting if the users did set one.

> e.g. would it potentially be possible to completely remove format_dict in IterableDataset.iter in the case that ex_iterable.iter_arrow evaluates to False, since in this case extraction+formatting has presumably already occurred?

the output of map can be anything in MappedExamplesIterable, it's the IterableDataset's job to format it if the user asks for a specific format. If there is no format set by the user, we yield the example as is (the data can even contain arbitrary python objects). E.g. if a user wants to do some processing from numpy arrays and return arbitrary python objects they should do `ds = ds.with_format("np").map(fn).with_format(None)`

So no I don't think we can remove the format_dict in IterableDataset (or we would have to add FormattedExamplesIterable after each MappedExamplesIterable but this doesn't allow disabling formatting after a map)

> @lhoestq what is the intended chaining logic? Should with_format.map.with_format override the original with_format and change the formatting of the inputs to map? Or is it intended to apply to the outputs?

the second with_format sets the format of the data yielded when iterating on the IterableDataset - it doens't change the input to map

Reply:
so, just to be sure, the default behaviour of with_format().map() should be to format both the inputs and the outputs of map?

only asking because it looked like for the non-iterable Dataset.map formatting only applies to the inputs but wasn't sure I was understanding that correctly.

Reply:
the output of Dataset.map is written as Arrow data on disk, but when the data is accessed it's formatted using the Dataset format type (map doesn't change the format of the dataset)

that's why a code likes this still returns numpy arrays:
```python
>>> ds = with_format("np").map(...)
>>> ds[0]  # dict of numpy arrays
```
and it can be reset to python objects with
```python
>>> ds = with_format("np").map(...).with_format("python")
>>> ds[0]  # dict of python objects
```

Reply:
ok got it thanks!